File: scala3_mutex/src/main/scala/org/chelck/dp/Table.scala
package org.chelck.dp

import scala.collection.mutable

class Table(val numberOfPhilosophers: Int) {
  private var usedForks = mutable.Set[Int]()
  private var philosophersWithForks = mutable.Map[Int, mutable.Set[Int]]()
  (0 until numberOfPhilosophers).foreach((p: Int) => this.philosophersWithForks.put(p, mutable.Set[Int]()))

  def getLeftFork(philosopher: Int): Option[Int] = {
    assert(philosopher >= 0 && philosopher < numberOfPhilosophers, "philosopher is " + philosopher)
    val leftFork: Int = philosopher
    //println(s"${philosopher}: ${leftFork}, UsedForks: ${usedForks}, contains: ${usedForks.contains(leftFork)}, isLast: ${isLastPhilosopher("Use left fork", philosopher)}")

    if (usedForks.contains(leftFork) || isLastPhilosopher("Use left fork", philosopher)) None
    else {
      usedForks.add(leftFork)
      philosophersWithForks(philosopher).add(leftFork)
      isLastPhilosopher("Use left fork", philosopher)
      Some(leftFork)
    }
  }

  def getRightFork(philosopher: Int): Option[Int] = {
    assert(philosopher >= 0 && philosopher < numberOfPhilosophers, "philosopher is " + philosopher)
    val rightFork: Int = (philosopher + 1) % numberOfPhilosophers
    //println(s"${philosopher}: getRighttFork: ${rightFork}, UsedForks: ${usedForks}, contains: ${usedForks.contains(rightFork)}, isLast: ${isLastPhilosopher("Use right fork", philosopher)}")

    if (usedForks.contains(rightFork) || isLastPhilosopher("Use right fork", philosopher)) None
    else {
      usedForks.add(rightFork)
      philosophersWithForks(philosopher).add(rightFork)
      isLastPhilosopher("Use right fork", philosopher)
      Some(rightFork)
    }
  }

  def returnLeftFork(philosopher: Int): Unit = {
    assert(philosopher >= 0 && philosopher < numberOfPhilosophers, "philosopher is " + philosopher)
    val leftFork: Int = philosopher

    usedForks.remove(leftFork)
    philosophersWithForks(philosopher).remove(leftFork)
    //println(s"${philosopher}: returnLeftFork: ${leftFork}, UsedForks: ${usedForks}, contains: ${usedForks.contains(leftFork)}")
  }

  def returnRightFork(philosopher: Int): Unit = {
    assert(philosopher >= 0 && philosopher < numberOfPhilosophers, "philosopher is " + philosopher)
    val rightFork: Int = (philosopher + 1) % numberOfPhilosophers
    usedForks.remove(rightFork)
    philosophersWithForks(philosopher).remove(rightFork)
    //println(s"${philosopher}: returnRightFork: ${rightFork}, UsedForks: ${usedForks}, contains: ${usedForks.contains(rightFork)}")
  }

  private def isLastPhilosopher(comment: String, philosopher: Int): Boolean = {

    val philosophersWithNoForks = for {
      (philosopherId, forks) <- this.philosophersWithForks
      if forks.isEmpty
    }
      yield philosopherId

    //println(s"comment: ${comment}, no forks: ${philosophersWithNoForks}")
    philosophersWithNoForks.size == 1 && philosophersWithNoForks.count(_ == philosopher) == 1
  }
}

File: java17/src/main/java/org/chelck/dp/Table.java
package org.chelck.dp;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Table {
    private final int numberOfPhilosophers;
    private final Set<Integer> usedForks;
    private final Map<Integer, Set<Integer>> philosophersWithForks;

    public Table(int numberOfPhilosophers) {
        this.numberOfPhilosophers = Objects.requireNonNull(numberOfPhilosophers);
        this.usedForks = new HashSet<>();
        this.philosophersWithForks = new HashMap<>();

        IntStream.range(0, numberOfPhilosophers).forEach(p -> this.philosophersWithForks.put(p, new HashSet<>()));
    }

    public Optional<Integer> getLeftFork(int philosopher) {
        assert philosopher >= 0 && philosopher < numberOfPhilosophers : "philosopher is " + philosopher;
        int leftFork = philosopher;

        if (usedForks.contains(leftFork) || isLastPhilosopher("Use left fork", philosopher)) {
            return Optional.empty();
        }
        else {
            usedForks.add(leftFork);
            philosophersWithForks.get(philosopher).add(leftFork);
            isLastPhilosopher("Use left fork", philosopher);
            return Optional.of(leftFork);
        }
    }

    public Optional<Integer> getRightFork(int philosopher) {
        assert philosopher >= 0 && philosopher < numberOfPhilosophers : "philosopher is " + philosopher;
        int rightFork = (philosopher + 1) % numberOfPhilosophers;

        if (usedForks.contains(rightFork) || isLastPhilosopher("Use right fork", philosopher)) {
            return Optional.empty();
        }
        else {
            usedForks.add(rightFork);
            philosophersWithForks.get(philosopher).add(rightFork);
            isLastPhilosopher("Use right fork", philosopher);
            return Optional.of(rightFork);
        }
    }

    public void returnLeftFork(int philosopher) {
        assert philosopher >= 0 && philosopher < numberOfPhilosophers : "philosopher is " + philosopher;

        int leftFork = philosopher;
        usedForks.remove(leftFork);
        philosophersWithForks.get(philosopher).remove(leftFork);
    }

    public void returnRightFork(int philosopher) {
        assert philosopher >= 0 && philosopher < numberOfPhilosophers : "philosopher is " + philosopher;
        int rightFork = (philosopher + 1) % numberOfPhilosophers;
        usedForks.remove(rightFork);
        philosophersWithForks.get(philosopher).remove(rightFork);
    }

    private boolean isLastPhilosopher(String comment, int philosopher) {
        List<Integer> philosophersWithNoForks = this.philosophersWithForks.entrySet().stream()
                .filter(e -> e.getValue().isEmpty())
                .map(e -> e.getKey())
                .collect(Collectors.toList());

        //System.out.println("CJH no forks: " + philosophersWithNoForks);

        return philosophersWithNoForks.size() == 1 && philosophersWithNoForks.contains(philosopher);    }
}
